src/goto-instrument/contracts/contracts.cpp

```diff
@@ -311,137 +330,157 @@ void code_contractst::check_apply_loop_contracts(
-
-  // Add the havocing code, but only check against the enclosing scope's
-  // write set if it was manually specified.
-  if(assigns_clause.is_nil())
-    insert_before_swap_and_advance(
-      goto_function.body,
-      loop_head,
-      add_pragma_disable_assigns_check(generated_code));
-  else
-    insert_before_swap_and_advance(
-      goto_function.body, loop_head, generated_code);
-
-  // ^^^ FIXME ^^^
-  // comment by delmasrd: I did not reactivate this behaviour
-  // because I think we always want to check the loop assignments against
-  // the surrounding clause
+  havoc_gen.append_full_havoc_code(loop_head_location, pre_loop_head_instrs);
 
-  insert_before_swap_and_advance(goto_function.body, loop_head, generated_code);
+  // Insert the second block of pre_loop_head_instrs: the havocing code.
+  // We do not `add_pragma_disable_assigns_check`,
+  // so that the enclosing scope's assigns clause instrumentation
+  // would pick these havocs up for inclusion (subset) checks.
+  goto_function.body.destructive_insert(loop_head, pre_loop_head_instrs);
 
   // Generate: assume(invariant) just after havocing
-  // We use a block scope to create a temporary assumption,
-  // and immediately convert it to goto instructions.
+  // We use a block scope for assumption, since it's immediately goto converted,
+  // and doesn't need to be kept around.
   {
-    code_assumet assumption{invariant_expr()};
-    assumption.add_source_location() = loop_head->source_location();
-    converter.goto_convert(assumption, generated_code, mode);
+    code_assumet assumption{invariant};
+    assumption.add_source_location() = loop_head_location;
+    converter.goto_convert(assumption, pre_loop_head_instrs, mode);
   }
 
   // Create fresh temporary variables that store the multidimensional
   // decreases clause's value before and after the loop
   for(const auto &clause : decreases_clause.operands())
   {
     const auto old_decreases_var =
-      new_tmp_symbol(
-        clause.type(), loop_head->source_location(), mode, symbol_table)
+      new_tmp_symbol(clause.type(), loop_head_location, mode, symbol_table)
         .symbol_expr();
-    generated_code.add(goto_programt::make_decl(
-      old_decreases_var, loop_head->source_location()));
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(old_decreases_var, loop_head_location));
     old_decreases_vars.push_back(old_decreases_var);
 
     const auto new_decreases_var =
-      new_tmp_symbol(
-        clause.type(), loop_head->source_location(), mode, symbol_table)
+      new_tmp_symbol(clause.type(), loop_head_location, mode, symbol_table)
         .symbol_expr();
-    generated_code.add(goto_programt::make_decl(
-      new_decreases_var, loop_head->source_location()));
+    pre_loop_head_instrs.add(
+      goto_programt::make_decl(new_decreases_var, loop_head_location));
     new_decreases_vars.push_back(new_decreases_var);
   }
 
   // TODO: Fix loop contract handling for do/while loops.
```

Comment:
Do we track this somewhere? We are using "FIXME" and "TODO" as labels for such comments. I'd pick one and make consistent across this file.